File: brackets.js
import * as readline from 'readline';

const opening = "{[(<";
const ending = "}])>";

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function main(chain) {
    let openings = 0;
    const index = [];
    let seen = 0;

    for (const c of chain) {
        seen++;

        const c_op = opening.includes(c);
        const c_ed = ending.includes(c);

        if (c_op) {
            openings++;
            index.push(opening.indexOf(c));
        }

        if (c_ed) {
            if (openings < 1 || index[index.length - 1] !== ending.indexOf(c)) {
                console.log("Bad!!");
                rl.close();
                return;
            }
            openings--;
            index.pop();
        }

        console.log(`${c} ${(100 * seen / chain.length).toFixed(2)}%`);
    }

    if (openings !== 0) {
        console.log("Bad!!");
    } else {
        console.log("Good!!");
    }

    rl.close();
}

rl.question('Your code in here: ', main);

File: brackets.py
OPENING: str = "{[(<"
ENDING: str = "}])>"

chain: str = input("Your code in here: ")

openings: int = 0
index: list = []

seen: int = 0
for c in chain:
    seen += 1

    c_op: bool = c in OPENING
    c_ed: bool = c in ENDING

    if c_op:
        openings += 1
        index.append(OPENING.find(c))

    if c_ed:
        if openings < 1 or index[-1] != ENDING.find(c):
            print("Bad!!")
            exit()
        openings -= 1
        index.pop()

    print(f"{c} {100*seen/len(chain)}%")

if openings != 0:
    print("Bad!!")
    exit()

print("Good!!")
